pkg/agent/cache/manager.go

```diff
@@ -0,0 +1,387 @@
+package cache
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+
+	"crypto/tls"
+	"crypto/x509"
+	"github.com/sirupsen/logrus"
+	spiffe_tls "github.com/spiffe/go-spiffe/tls"
+	"github.com/spiffe/go-spiffe/uri"
+	"github.com/spiffe/spire/pkg/common/util"
+	"github.com/spiffe/spire/proto/api/node"
+	proto "github.com/spiffe/spire/proto/common"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"io"
+	"sync"
+	"time"
+)
+
+type Manager interface {
+	Init()
+	Cache() Cache
+	Shutdown(err error)
+	Done() <-chan struct{}
+	Err() error
+}
+
+type EntryRequest struct {
+	CSR   []byte
+	entry CacheEntry
+}
+
+type MgrConfig struct {
+	ServerCerts    []*x509.Certificate
+	ServerSPIFFEID string
+	ServerAddr     string
+	BaseSVID       []byte
+	BaseSVIDKey    *ecdsa.PrivateKey
+	BaseRegEntries []*proto.RegistrationEntry
+	Logger         logrus.FieldLogger
+}
+
+type manager struct {
+	managedCache     Cache
+	serverCerts      []*x509.Certificate
+	serverAddr       string
+	serverSPIFFEID   string
+	reason           error
+	entryRequestCh   chan map[string][]EntryRequest
+	regEntriesCh     chan []*proto.RegistrationEntry
+	cacheEntryCh     chan CacheEntry
+	spiffeIdEntryMap map[string]CacheEntry
+	baseSVID         []byte
+	baseSVIDKey      *ecdsa.PrivateKey
+	baseSPIFFEID     string
+	baseRegEntries   []*proto.RegistrationEntry
+	log              logrus.FieldLogger
+	ctx              context.Context
+	cancel           context.CancelFunc
+	once             sync.Once
+	doneCh           chan struct{}
+}
+
+func NewManager(ctx context.Context, c *MgrConfig) (Manager, error) {
+
+	cert, err := x509.ParseCertificate(c.BaseSVID)
+	if err != nil {
+		return nil, err
+	}
+	basespiffeID, err := uri.GetURINamesFromCertificate(cert)
+	if err != nil {
+		return nil, err
+	}
+	ctx, cancel := context.WithCancel(ctx)
+
+	return &manager{
+		managedCache:     NewCache(c.Logger),
+		serverCerts:      c.ServerCerts,
+		serverSPIFFEID:   c.ServerSPIFFEID,
+		serverAddr:       c.ServerAddr,
+		baseSVID:         c.BaseSVID,
+		baseSVIDKey:      c.BaseSVIDKey,
+		baseSPIFFEID:     basespiffeID[0],
+		baseRegEntries:   c.BaseRegEntries,
+		log:              c.Logger.WithField("subsystem_name", "cacheManager"),
+		spiffeIdEntryMap: make(map[string]CacheEntry),
+		entryRequestCh:   make(chan map[string][]EntryRequest),
+		regEntriesCh:     make(chan []*proto.RegistrationEntry),
+		cacheEntryCh:     make(chan CacheEntry),
+		ctx:              ctx,
+		cancel:           cancel,
+		doneCh:           make(chan struct{}),
+	}, nil
+}
+
+func (m *manager) Shutdown(err error) {
+	m.once.Do(func() {
+		m.reason = err
+	})
+	m.cancel()
+}
+
+func (m *manager) Done() <-chan struct{} {
+	return m.doneCh
+}
+func (m *manager) Err() error {
+	<-m.Done()
```

Comment:
Above near my other comment, we wait on the Done channel in agent, then once we receive off of it, we call Err() to see if there was an error... but Done channel is being read here again. Does it get closed somewhere? If it's closed we'll read a zero value and continue, but if it's not it will block

Reply:
yes it gets closed when the main loop returns..see comment above